### ISSUE TITLE
FSDP use_orig_params does not expose original parameters

### ISSUE THREAD

**Issue by Shinechaote** [pytorch/pytorch#136314]
### 🐛 Describe the bug

I am trying to use FSDP to train a model and I need the shape of the parameters to be 2D for the optimizer I am using. FSDP however flattens them even if I use use_orig_params=True

Expected output of the code I provided below:E
```
fc1.weight 2 torch.Size([20, 20])
fc1.weight 2 torch.Size([20, 20])
```

Actual output when running ```torchrun --nproc_per_node=2 minimal.py ```:
```
fc1.weight 2 torch.Size([20, 20])
fc1.weight 1 torch.Size([400])
```I

I get the correct output when running on only 1 node.

```
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP

# A simple model with 2D weights (Linear layer)
class SimpleModel(nn.Module):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.fc1 = nn.Linear(20, 20)  # 2D weight: (2, 4)
        self.fc2 = nn.Linear(20, 20)  # 2D weight: (2, 4)

    def forward(self, x):
        x = self.fc1(x)
        x = self.fc2(x)
        return x

def setup_distributed(backend='nccl'):
    # Initialize the process group
    dist.init_process_group(backend)

def cleanup_distributed():
    # Clean up the process group
    dist.destroy_process_group()

def main():
    # Initialize distributed environment
    setup_distributed()
    torch.cuda.set_device(torch.distributed.get_rank())

    # Create the model and move it to GPU
    model = SimpleModel().cuda()

    if torch.cuda.current_device() == 0:
        for n, p in model.named_parameters():
            print(n, p.ndim, p.shape) 
            break
        
    # Wrap the model with FSDP to shard its parameters
    model = FSDP(model, use_orig_params=True)
    
    if torch.cuda.current_device() == 0:
        for n, p in model.module.named_parameters():
            print(n, p.ndim, p.shape)
            break

    # Clean up the distributed environment
    cleanup_distributed()

if __name__ == "__main__":
    main()

``` 

### Versions

PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-513.5.1.el8_9.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB

Nvidia driver version: 555.42.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.0.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7313 16-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 1
Core(s) per socket:                 16
Socket(s):                          2
Stepping:                           1
BogoMIPS:                           5989.01
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          1 MiB (32 instances)
L1i cache:                          1 MiB (32 instances)
L2 cache:                           16 MiB (32 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       8
NUMA node0 CPU(s):                  0-3
NUMA node1 CPU(s):                  4-7
NUMA node2 CPU(s):                  8-11
NUMA node3 CPU(s):                  12-15
NUMA node4 CPU(s):                  16-19
NUMA node5 CPU(s):                  20-23
NUMA node6 CPU(s):                  24-27
NUMA node7 CPU(s):                  28-31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] galore-torch==1.0
[pip3] json-numpy==2.1.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.4.1
[pip3] torchaudio==2.2.0
[pip3] torchvision==0.17.0
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] galore-torch              1.0                      pypi_0    pypi
[conda] json-numpy                2.1.0                    pypi_0    pypi
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.10          py310h5eee18b_0  
[conda] mkl_random                1.2.7           py310h1128e8f_0  
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-cuda              12.4                 hc786d27_6    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     2.4.1                    pypi_0    pypi
[conda] torchaudio                2.2.0                    pypi_0    pypi
[conda] torchvision               0.17.0                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

**Comment by janEbert** [comment]
You need to use the [`model.summon_full_params`](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.summon_full_params) context to obtain the "full" parameters, i.e., the unsharded versions.

**Comment by Shinechaote** [comment]
I get the expected output using your answer but I dont think it is the solution I am looking for. 

The solution just puts all the parameters on every gpu as far as I can tell but this will not be feasible for the model I am trying to use. Also the solution works without setting use_orig_params=True, what is the purpose of this parameter then?

**Comment by janEbert** [comment]
Maybe you can summon individual parameters to reduce the memory requirements?
For me, IIRC, I had to use `use_orig_params=True` at one point to enable per-parameter learning rates, which was not supported with `use_orig_params=False`.

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/blob/65df26f61559ec580b8e455c8833b19358298cc2/torch/distributed/fsdp/fully_sharded_data_parallel.py#L375-L389

`use_orig_params=True` just exposes out views of the original parameters into the underlying 1D `FlatParameter`. We cannot ensure that it preserves tensor structure since we are following flat-parameter sharding (flatten, concat, chunk a list of parameters).

If you want per-parameter sharding, FSDP2 does so directly on dim-0. You will see a DTensor with `Shard(0)` placement on the data parallel mesh dimension. https://github.com/pytorch/pytorch/issues/114299
You may need to do some work to have your optimizer work with DTensor, but this should be the more promising direction. If you have a custom optimizer and making it work with DTensor seems too challenging, then (1) please let us know about this, and (2) feel free to convert the DTensor to `torch.Tensor` via `DTensor.to_local()`.
